### PR TITLE
[Merged by Bors] - feat: port data.pfun

### PR DESCRIPTION
**Pull request by arienmalec**
Port of data.pfun

### CODE REVIEW

**Comment by arienmalec** [comment]
Did all the easy fixes and doc fixes -- still needs lots of tatics fixes

File: Mathlib/Data/PFun.lean [hunk 1]
```diff
@@ -0,0 +1,714 @@
...
+open Function
+
+/-- `PFun α β`, or `α →. β`, is the type of partial functions from
+  `α` to `β`. It is defined as `α → part β`. -/
+def PFun (α β : Type _) :=
+  α → Part β
+#align pfun PFun
+
+-- mathport name: «expr →. »
+/-- Terminology expressing `PFun α β` as `α  →. β` -/
```
Comment (jcommelin) [inline Mathlib/Data/PFun.lean:67]:
```suggestion
/-- `α  →. β` is notation for the type `PFun α β` of partial functions from `α` to `β`.  -/
```

**Review by jcommelin — COMMENTED** [review]
Why do we need to add `noncomputable` to several decls? I think `PFun` is a type where we might want to care about computability.

**Comment by Komyyy** [comment]
Data/PFun.lean includes functions related to `Acc`, like `PFun.fix` and `PFun.fixInduction`.
In Lean3, the `noncomputable` kw was not necessary to decls related to `Acc`, but in Lean4, the compiler still can't compile a function with `Acc.rec` to bytecode. However, the kernel can evaluate a function with `Acc.rec`, so the `noncomputable` kw is newly required in Lean4, like `WellFounded.fix`.
Additionally, even a function with `Acc.rec`, can become computable with `extern`, like `Nat.mod`.

File: Mathlib/Data/PFun.lean [hunk 2]
```diff
@@ -0,0 +1,714 @@
...
+#align pfun.prod_map_id_id PFun.prodMap_id_id
+
+@[simp]
+theorem prodMap_comp_comp (f₁ : α →. β) (f₂ : β →. γ) (g₁ : δ →. ε) (g₂ : ε →. ι) :
+    (f₂.comp f₁).prodMap (g₂.comp g₁) = (f₂.prodMap g₂).comp (f₁.prodMap g₁) := -- by
+  -- Porting note: was `by tidy`, below is a golf'd verson of the `tidy?` proof
+  ext $ λ ⟨_, _⟩ ⟨_, _⟩ =>
+  ⟨λ ⟨⟨⟨h1l1, h1l2⟩, ⟨h1r1, h1r2⟩⟩, h2⟩ => ⟨⟨⟨h1l1, h1r1⟩, ⟨h1l2, h1r2⟩⟩, h2⟩,
+   λ ⟨⟨⟨h1l1, h1r1⟩, ⟨h1l2, h1r2⟩⟩, h2⟩ => ⟨⟨⟨h1l1, h1l2⟩, ⟨h1r1, h1r2⟩⟩, h2⟩⟩
+
```
Comment (ChrisHughes24) [inline Mathlib/Data/PFun.lean:711]:
```suggestion
```

**Comment by ChrisHughes24** [comment]
bors merge

**Pull request closed without merging**